nodes/filter/single.go

```diff
@@ -1,27 +1,345 @@
 package filter
 
 import (
+	"fmt"
+	"math"
+
 	"github.com/NubeDev/flow-eng/node"
 )
 
-type OnlyTrue struct {
+// type OnlyTrue struct {
+// 	*node.Spec
+// }
+
+// type OnlyFalse struct {
+// 	*node.Spec
+// }
+
+type PreventNull struct {
+	*node.Spec
+	lastValue bool
+}
+
+type PreventEqualFloat struct {
+	*node.Spec
+	lastValue float64
+}
+
+type PreventEqualString struct {
+	*node.Spec
+	lastValue string
+}
+
+type OnlyEqualFloat struct {
+	*node.Spec
+	lastValue float64
+}
+
+type OnlyEqualString struct {
+	*node.Spec
+	lastValue string
+}
+
+type OnlyBetween struct {
+	*node.Spec
+	lastValue float64
+}
+
+type OnlyGreater struct {
+	*node.Spec
+	lastValue float64
+}
+
+type OnlyLower struct {
+	*node.Spec
+	lastValue float64
+}
+
+type PreventDuplicates struct {
 	*node.Spec
+	lastValue float64
 }
 
-func NewOnlyTrue(body *node.Spec) (node.Node, error) {
-	body = node.Defaults(body, onlyTrue, category)
+// func NewOnlyTrue(body *node.Spec) (node.Node, error) {
+// 	body = node.Defaults(body, onlyTrue, category)
+// 	inputs := node.BuildInputs(node.BuildInput(node.In, node.TypeBool, nil, body.Inputs))
+// 	outputs := node.BuildOutputs(node.BuildOutput(node.Out, node.TypeBool, nil, body.Outputs))
+// 	body = node.BuildNode(body, inputs, outputs, nil)
+// 	body.SetHelp(fmt.Sprintln("This node filters ‘input’ values. Only ‘true’ values are passed to ‘output’."))
+// 	return &OnlyTrue{body}, nil
+// }
+
+// func NewOnlyFalse(body *node.Spec) (node.Node, error) {
+// 	body = node.Defaults(body, onlyFalse, category)
+// 	inputs := node.BuildInputs(node.BuildInput(node.In, node.TypeBool, nil, body.Inputs))
+// 	outputs := node.BuildOutputs(node.BuildOutput(node.Out, node.TypeBool, nil, body.Outputs))
+// 	body = node.BuildNode(body, inputs, outputs, nil)
+// 	body.SetHelp(fmt.Sprintln("This node filters ‘input’ values. Only ‘false’ values are passed to ‘output’."))
+// 	return &OnlyFalse{body}, nil
+// }
+
+func NewPreventNull(body *node.Spec) (node.Node, error) {
+	body = node.Defaults(body, preventNull, category)
 	inputs := node.BuildInputs(node.BuildInput(node.In, node.TypeBool, nil, body.Inputs))
 	outputs := node.BuildOutputs(node.BuildOutput(node.Out, node.TypeBool, nil, body.Outputs))
 	body = node.BuildNode(body, inputs, outputs, nil)
-	return &OnlyTrue{body}, nil
+	body.SetHelp(fmt.Sprintln("This node filters ‘input’ values. All ‘input’ values are passed to ‘output’ EXCEPT ‘null’. Output the last known value if the input became null."))
+	return &PreventNull{body, false}, nil
+}
+
+func NewPreventEqualFloat(body *node.Spec) (node.Node, error) {
+	body = node.Defaults(body, preventEqualFloat, category)
+	inputs := node.BuildInputs(node.BuildInput(node.In, node.TypeFloat, nil, body.Inputs), node.BuildInput(node.In1, node.TypeFloat, nil, body.Inputs))
```

Comment:
Please rename to match input/output names from the knowledge base: https://nubeio.zohodesk.com.au/portal/en/kb/articles/rubix-wires-filter-nodes#Settings_4